### PR TITLE
Make root-ish definition static

### PR DESCRIPTION
**Pull request by fjetter**
@gjoseph92 The topic of dynamic root task definition were brought up quite frequently in the past. I never figured out why the cluster size has to be part of the classification logic.
From everything I understand this should be moved to the actual decision logic and not the classification. The property of being rootish, the way it is defined right now, is actually a static _TaskGroup_ property.

In the case of co-assignment, I can understand that we do not want to assign too small task groups but why not handle this case in the decision logic itself instead of the classification?

TODO
- [ ] Fix two failing tests since they appear to be related
- [ ] Clean up test cases

### CODE REVIEW

File: distributed/scheduler.py [hunk 1]
```diff
@@ -2052,6 +2052,7 @@ def decide_worker_rootish_queuing_disabled(
             and tg.last_worker_tasks_left
             and lws.status == Status.running
             and self.workers.get(lws.address) is lws
+            and len(tg) > self.total_nthreads * 2
```
Comment (fjetter) [inline distributed/scheduler.py:2055]: Putting this statement here makes it also much easier to motivate it.

Below, we're calculating the number of tasks of the given task group we can assign safely to a worker `(len(tg) / self.total_nthreads) * ws.nthreads`

Therefore, `last_worker_tasks_left > 2 * ws.nthreads ~ tasks_assigned / thread > 2`

This means, we're only scheduling using co-assignment if we can at least assign two tasks per thread. Otherwise, we abort to protect our ability to scale out and use the ordinary worker_objective to distribute tasks

Comment (fjetter) [reply distributed/scheduler.py:2055]: Whether or not this "at least two tasks per thread" logic is a good choice is an entirely different matter, of course. 
However, this gives some meaning to this otherwise arbitrary choice

**Comment by fjetter** [comment]
My first pass over failing tests didn't reveal anything scandalous.
- test_priorities file put in some test assumptions about how many tasks are supposed to be queued on scheduler / worker when a worker is paused. This counting obviously changes with queuing. I also had to set the saturation to exactly 1.0 for these tests to even work the way they are. We may need to review these changes more carefully
- A couple of tests that just don't make sense if any of the submitted tasks are treated as being queued. I set the worker saturation to inf in these cases. They mostly test worker logic and this pick is fine imo
- The most interesting failure that is entirely unrelated to queuing is `test_scatter_compute_store_lose_processing`. There is a race condition on worker shutdown that causes a client never to receive cancelled-keys messages. This is only popping up here now because queuing slightly changes the timing of the worker close (since w/ queuing we don't need to wait for the worker threadpool to finish)

File: distributed/scheduler.py [hunk 2]
```diff
@@ -2819,15 +2820,15 @@ def is_rootish(self, ts: TaskState) -> bool:
...
-        if ts.resource_restrictions or ts.worker_restrictions or ts.host_restrictions:
+        if (
+            ts.resource_restrictions
+            or ts.worker_restrictions
+            or ts.host_restrictions
+            or ts.actor
+        ):
             return False
         tg = ts.group
-        # TODO short-circuit to True if `not ts.dependencies`?
```
Comment (fjetter) [inline distributed/scheduler.py:2825]: If this all works out well, I would actually like to raise this method to a `TaskGroup` property instead. We can implement a similar short-circuit there although I'm not sure if this is actually worth it.

File: distributed/scheduler.py [hunk 3]
```diff
@@ -2074,7 +2075,16 @@ def decide_worker_rootish_queuing_disabled(
...
 
-    def decide_worker_rootish_queuing_enabled(self) -> WorkerState | None:
+    def worker_objective_rootish_queuing(self, ws, ts):
+        # FIXME: This is basically the ordinary worker_objective but with task
+        # counts instead of occupancy.
+        comm_bytes = sum(
+            dts.get_nbytes() for dts in ts.dependencies if ws not in dts.who_has
+        )
+        # See test_nbytes_determines_worker
+        return (len(ws.processing) / ws.nthreads, comm_bytes, ws.nbytes)
```
Comment (fjetter) [inline distributed/scheduler.py:2085]: @gjoseph92 I remember you very strongly opposing adding an objective function that respect comm_bytes. I remember vaguely a point about widely shared dependencies but I do not see an issue with this.

We need an objective like this to ensure very basic scheduling policies to be respected, e.g. `test_nbytes_determines_worker` where a dependent task is supposed to be scheduled on a worker holding most of its dependencies.

This should only truly matter if there are multiple workers available with a similar number of free slots. My intuition tells me this objective function should still make sense

Comment (fjetter) [reply distributed/scheduler.py:2085]: xref https://github.com/dask/distributed/pull/7280

**Comment by fjetter** [comment]
there is a likely related failure `distributed/tests/test_scheduler.py::test_deadlock_resubmit_queued_tasks_fast` on ubu3.9 / no-queue I need to check out

**Comment by fjetter** [comment]
benchmarks are mostly harmless for this commit. we get a nice kick for some array slicing operations both in runtime and peak memory (plot shows peak)

image

https://github.com/coiled/coiled-runtime/actions/runs/4136037002

File: distributed/scheduler.py (hunk 3 above)
Comment (gjoseph92) [reply distributed/scheduler.py:2085]: > `test_nbytes_determines_worker` where a dependent task is supposed to be scheduled on a worker holding most of its dependencies

**tl;dr** then that task shouldn't have been queued in the first place, and the root-ish heuristic is at fault, not the objective function. For a task to be queued, it must be equally good to run on any worker.

I do oppose an objective function that considers data transfer _for root-ish tasks_. It contradicts the definition of root-ish-ness: that the location of the tasks' dependencies are not important for scheduling purposes.

When there's only one thread open in the cluster, on a random worker, the next task on the queue must be equally "good" to run on that random worker as on any other worker in the cluster. If it would have been "better" if the thread had opened on a different particular worker—because that worker had more dependencies in memory, say, or because it had some resource—then the task should not have been on the queue in the first place. (This is why putting tasks with restrictions on the queue breaks the core assumption of queuing: https://github.com/dask/distributed/pull/7526.)

This is important because we choose the pool of candidate workers for queued tasks based only on how busy they are. That works because we can guarantee there are no other factors that should play into worker selection for these tasks.

The root-ish-ness heuristic is meant to provide that guarantee, by identifying task groups where:
1. The tasks will end up being assigned to every worker in the cluster (`len > nthreads`)
2. Therefore, any common dependencies the tasks have will end up being replicated onto every worker in the cluster (`deps len < 5`)

The key is the second point: because the dependencies will be replicated to all workers, we don't need to select workers that avoid data transfer—in the long term, the data transfer is unavoidable. Indeed, we even know of some cases (that don't meet the root-ish heuristic, to be clear) where trying to avoid a transfer leads to worse scheduling choices: https://github.com/dask/distributed/issues/6570.

So the reason I oppose an objective function that considers data transfer for root-ish tasks is because it _implies_ that one worker might be preferable over another based on something besides busyness, which violates these core assumptions.

But to be clear: I only oppose this on the grounds of logical consistency. I don't expect an objective function that considers transfer will actually make a measurable different in most cases. (In most cases, the size of the candidate pool is 1 anyway.) But since it contradicts this key assumption of queuing, I'd want to come up with logical proof of why it's safe—and why it's preferable—especially since we know this type of weighting leads to bad decisions in other cases (https://github.com/dask/distributed/issues/6570).

I haven't bothered to put in the effort to make the proof, since I haven't seen what the gain would be. I'm happy to be shown that it's both safe and better, though—again, I just haven't seen a reason to.

See also:
- https://github.com/dask/distributed/pull/7280#discussion_r1018608580
- https://github.com/dask/distributed/pull/7280#discussion_r1019420197

File: distributed/scheduler.py [hunk 4]
```diff
@@ -2819,15 +2829,15 @@ def is_rootish(self, ts: TaskState) -> bool:
...
+        ):
             return False
         tg = ts.group
-        # TODO short-circuit to True if `not ts.dependencies`?
-        return (
-            len(tg) > self.total_nthreads * 2
-            and len(tg.dependencies) < 5
-            and sum(map(len, tg.dependencies)) < 5
-        )
+        return len(tg.dependencies) < 5 and sum(map(len, tg.dependencies)) < 5
```
Comment (gjoseph92) [inline distributed/scheduler.py:2840]: I'm concerned about losing the `len(tg) > self.total_nthreads` check. That means that, for example, task `c` would now be considered root-ish:
```
  c
 / \
a   b
```

We don't want task `c` to be queued, for reasons described in the comment above. `c` isn't equally good to run on any worker—it should only be run on the workers that hold `a` or `b`. It's better for `c` run on the right worker, but maybe have to wait, than run on whichever wrong worker has a task slot open first.

The point of the `total_nthreads` check is that it identifies fan-out situations, where we go from dependencies that are on very few workers to dependents that will be on every worker. The fact that the dependencies will be copied to every worker is what allows us to ignore them while scheduling, which is what allows us to queue the tasks, since in queueing, worker busyness must be the only important scheduling factor.

**Review by gjoseph92 — COMMENTED** [review]
I also really want to make the root-ish definition static. However, I don't think we can drop the check against `total_nthreads` without changing some other fundamental things.

File: distributed/scheduler.py (hunk 4 above)
Comment (gjoseph92) [reply distributed/scheduler.py:2840]: Why did you add the check against `total_nthreads` to the `queuing_disabled` case, but not `queuing_enabled`? If we could add the check into `queuing_enabled` as well, then I think this would be fine. But I'm also not clear on what the benefit of all this change would be then—the definition still isn't "static" in a practical sense, we'd just be moving the non-static decision to a different point in the code.

File: distributed/scheduler.py (hunk 3 above)
Comment (fjetter) [reply distributed/scheduler.py:2085]: First of all, I think this is a moot point once I introduce some cutoff as suggested in the other comment. It would've been fine to just point out this error instead of writing such a long comment.

> It contradicts the definition of root-ish-ness: that the location of the tasks' dependencies are not important for scheduling purposes.

I disagree with this definition. The reason why we're having a special case for these tasks is because we assume they are data producers. The natural assumption is that they are at the bottom of our graph and are the most frequent source of data overproduction.
While I see some value for intermediate layers to be considered as root-ish I believe the same argument holds.
I don't think these perspectives are orthogonal but your definition is not the whole truth.

If something like https://github.com/dask/distributed/issues/6570 is truly the complexity driver here we should consider alternative approaches (inlining, better weights, scheduler hints, etc.)


> If it would have been "better" if the thread had opened on a different particular worker—because that worker had more dependencies in memory, say, or because it had some resource—then the task should not have been on the queue in the first place.

I also don't agree with this sentiment. I see a point for suppressing/ignoring dependencies but especially for cases where worker-saturation>1 I see how nuance can matter. Total randomness is not necessarily desired.

**Comment by fjetter** [comment]
I introduced a cutoff as suggested in https://github.com/dask/distributed/pull/7531#discussion_r1102728485 and elevated the rootish definition to a TG property.
Since this is now a graph/topological property we can also write proper tests. I copied a couple over from dask.order but do not intend to add much more there unless there is a good reason for it. I'm not entirely sure how valuable these are but I like having the minimal infrastructure in there in case we want to expand or modify the definition

File: distributed/scheduler.py [hunk 5]
```diff
@@ -1096,6 +1106,15 @@ def _to_dict_no_nest(self, *, exclude: Container[str] = ()) -> dict[str, Any]:
         """
         return recursive_to_dict(self, exclude=exclude, members=True)
 
+    @property
+    def rootish(self):
+        return (
+            not self._has_restrictions
+            and len(self) >= 5
+            and len(self.dependencies) < 5
+            and sum(map(len, self.dependencies)) < 5
```
Comment (gjoseph92) [inline distributed/scheduler.py:1115]:
```suggestion
            and (ls := len(self)) >= 5
            and len(self.dependencies) < 5
            and (ndeps := sum(map(len, self.dependencies))) < 5
            and ls / ndeps > 2
```
Maybe we're trying to get at the idea of "fanning out"? As in there are significantly (say 2x) more tasks in the group than there are inputs?

idk if this is valuable, just a thought. I think it would also eliminate the need for the magic `len(self) >= 5`.

**Review by gjoseph92 — COMMENTED** [review]
@fjetter this is looking better to me. I realize I think I'm also missing the larger picture here. What's the motivation for working on this right now? Is there some immediate next work that this would enable, or are you just looking to clean it up?

FWIW, making this definition static also closes https://github.com/dask/distributed/issues/7273.

File: distributed/scheduler.py (hunk 5 above)
Comment (fjetter) [reply distributed/scheduler.py]: I am a bit hesitant here because we don't have any examples (i.e. tests) that indicate the one or the other would be favorable. At least in the future we should motivate changes to this by new/modified tests.

Comment (fjetter) [reply distributed/scheduler.py]: If we went for your logic, why even include `len(self) >= 5`? Why not just

```python
(
            len(self.dependencies) < 5
            and (ndeps := sum(map(len, self.dependencies))) < 5
            # Fan-out
            and (len(self) / ndeps > 2 if ndeps else True)
        )
```

(FWIW this would also include the case of no deps naturally

**Review by fjetter — COMMENTED** [review]
> I realize I think I'm also missing the larger picture here. What's the motivation for working on this right now?

Mostly I was poking at the code base to get either a fix or a better understanding for the P2P problem (https://github.com/dask/distributed/issues/7496). Even after spending months with this topic, I was still surprised about the complexity that I decided to try a hard simplification.
I believe that maintenance and continued enhancements around this will otherwise not be feasible 
From a different perspective, if you feel the need to write comments like https://github.com/dask/distributed/pull/7531#discussion_r1101872001 even though tests are green and benchmarks look good, the code is too complex and/or not tested well enough.

File: distributed/scheduler.py (hunk 5 above)
Comment (fjetter) [reply distributed/scheduler.py]: What I don't like about this suggestion is that there is again another magic number `2` in there. Why would a fan-out only be a fan-out if there is a factor of 2? For entirely symmetric cases I can see how we can motivate 2 but not for asymmetric ones, e.g.

```mermaid
graph BT;
  A1-->B1;
  A1-->B2;
  A2-->B3;
  A2-->B4;
  A3-->B5;
```

Is this meaningfully different to the case where A3 splits into two tasks as well? Anyhow, I believe arguing about these small graphs is not truly relevant. As soon as we hit any real world graphs these magic numbers are getting more and more meaningless

Comment (gjoseph92) [reply distributed/scheduler.py]: > If we went for your logic, why even include `len(self) >= 5`?

Sorry, I meant to remove that check, just forgot to delete it in the diff. Agreed that it's unnecessary.

> For entirely symmetric cases I can see how we can motivate 2 but not for asymmetric ones, e.g.

This gets at a larger problem, which is the use of TaskGroups at all for these types of scheduling choices. We're assuming that TaskGroups indicate something structural about the graph, but they don't necessarily. I think we have to accept that we're assuming TaskGroups are:
* symmetrical
* correspond 1:1 with structural layers in the graph

If we're going to use TaskGroups for these sorts of heuristics, we have to be okay taking on those assumptions. I'd rather focus on graphs that meet these assumptions, but still would schedule poorly.

Comment (gjoseph92) [reply distributed/scheduler.py]: > I'm not even sure if "fanning out" is something we want to capture

I'm not sure either. Though fanning out is what the "ish" part of root-ish task logic was trying to capture. The idea was that when a tiny number of tasks "fan out" to a much wider number of tasks, if you don't copy the input tasks to most workers, you'll end up highly imbalanced, with most workers idle and a couple overwhelmed.

Mathematically though, I'm realizing that so long as we have `sum(map(len, self.dependencies)) < 5`, whether we call it fan-out, or a lower bound on `len(self)`, they're effectively equivalent. They're both expressing "there are very few tasks that go into the task group, and the task group is decently larger than the number of inputs". It's just arithmetic to turn one into the other.

> What I don't like about this suggestion is that there is again another magic number `2` in there

Yup, I don't like that either. I don't think we'll be able to come up with a _static_ definition for "should be queued" that doesn't include a magic number though, at least without significantly refactoring how the queued state works.

**Comment by fjetter** [comment]
Reran benchmarks with much more repetitions and everything looks OK. So, at least for the selection of workloads I ran this change likely does not impact the classification logic (as expected)

https://github.com/coiled/coiled-runtime/actions/runs/4172181281


<details>
<summary> Wall Clock </summary>

image


</details>

<details>
<summary> Avg memory </summary>

image


</details>


<details>
<summary> Peak memory </summary>

image


</details>
